File: cstuff/quicksort.c
#include <stdio.h>

void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int arr[], int low, int high) {
    int i, j, pivot;
    pivot = arr[high];
    i = low - 1;
    for (j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

void quicksort(int arr[], int low, int high) {
    if (low < high) {
        int j = partition(arr, low, high);
        quicksort(arr, low, j - 1);
        quicksort(arr, low + 1, high);
    }
}

void displayArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("arr[%d]: %d\n", i, arr[i]);
    }
}

int main() {
    int i, n;
    printf("Enter size of array!\n");
    printf("n: ");
    scanf("%d", &n);
    int arr[n];
    for (int i = 0; i < n; i++) {
        printf("arr[%d]: ", i);
        scanf("%d", &arr[i]);
    }

    printf("Current array is:\n");
    displayArray(arr, n);

    quicksort(arr, 0, n - 1);

    printf("Sorted array is:\n");
    displayArray(arr, n);
}

File: py/quicksort.py
#!/usr/bin/env python3

arr = []


def partition(low, high):
    pivot = arr[high]
    print(f"Pivot = {pivot} at index {high}")
    i = low - 1
    print(f"i = low - 1 = {i}")
    for j in range(low, high):
        if arr[j] < pivot:
            print(f"arr[{j}] = {arr[j]}")
            print(f"Incrementing i from {i} to {i + 1}")
            i = i + 1
            print(f"Swapping arr[i={i}] {arr[i]} and arr[j={j}] {arr[j]}")
            arr[i], arr[j] = arr[j], arr[i]
    print(f"Swapping arr[i+1={i+1}] {arr[i+1]} and arr[high={high}] {arr[high]}")
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    print(f"Returning i+1 = {i+1}")
    return i + 1


def quicksort(low, high):
    if low < high:
        print(f"low {low} is less than high {high}")
        j = partition(low, high)
        print(f"j is {j}")
        print(f"Sorting low = {low} to j-1 = {j-1}")
        quicksort(low, j - 1)
        print(f"Sorting j+1 = {j+1} to high = {high}")
        quicksort(j + 1, high)


for i in range(int(input("Enter size of array: "))):
    arr.append(int(input(f"Enter element {i}: ")))

print(arr)
quicksort(0, len(arr) - 1)
print(arr)
